exonum/src/storage/proof_map_index/key.rs

```diff
@@ -75,52 +69,117 @@ impl ::std::ops::Not for ChildKind {
-    pub fn to(&self) -> u16 {
-        self.to
+    /// Constructs the `ProofPath` from raw bytes.
+    fn from_raw(raw: [u8; PROOF_PATH_SIZE]) -> ProofPath {
+        debug_assert!(
+            if raw[PROOF_PATH_KIND_POS] == LEAF_KEY_PREFIX {
+                raw[PROOF_PATH_LEN_POS] == 0
+            } else {
+                true
+            },
+            "ProofPath is inconsistent"
+        );
+
+        ProofPath {
+            bytes: raw,
+            start: 0,
+        }
     }
 
-    /// Returns length of the `DBKey`.
-    pub fn len(&self) -> u16 {
-        self.to - self.from
+    /// Sets the right border of the bit range.
+    fn set_end(&mut self, end: u16) {
+        let max_len = (KEY_SIZE * 8) as u16;
+        assert!(end <= max_len);
+        // Update ProofPath kind and right bound.
+        if end == max_len {
+            self.bytes[0] = LEAF_KEY_PREFIX;
+            self.bytes[PROOF_PATH_LEN_POS] = 0;
+        } else {
+            self.bytes[0] = BRANCH_KEY_PREFIX;
+            self.bytes[PROOF_PATH_LEN_POS] = end as u8;
+        };
     }
+}
 
-    /// Returns true if `DBKey` has zero length.
-    pub fn is_empty(&self) -> bool {
-        self.to == self.from
+/// The bits representation of the `ProofPath`.
+pub trait BitsRange {
+    /// Returns the left border of the range.
+    fn start(&self) -> u16;
+    /// Returns the right border of the range.
+    fn end(&self) -> u16;
+    /// Returns length in bits of the range.
+    fn len(&self) -> u16 {
+        self.end() - self.start()
+    }
+    /// Returns true if the range has zero length.
+    fn is_empty(&self) -> bool {
+        self.end() == self.start()
     }
-
     /// Get bit at position `idx`.
-    pub fn get(&self, idx: u16) -> ChildKind {
-        debug_assert!(self.from + idx < self.to);
+    fn bit(&self, idx: u16) -> ChildKind;
+    /// Returns the new `ProofPath` with the given left border.
+    fn start_from(&self, idx: u16) -> Self;
+    /// Shortens this ProofPath to the specified length.
+    fn prefix(&self, pos: u16) -> Self;
+    /// Return object which represents a view on to this slice (further) offset by `pos` bits.
+    fn suffix(&self, pos: u16) -> Self;
+    /// Returns true if we starts with the same prefix at the whole of `other`
+    fn starts_with(&self, other: &Self) -> bool {
+        self.common_prefix(other) == other.len()
+    }
+    /// Returns how many bits at the beginning matches with `other`
+    fn common_prefix(&self, other: &Self) -> u16;
+    /// Returns the raw bytes of the key.
+    fn raw_key(&self) -> &[u8];
+}
+
+impl BitsRange for ProofPath {
+    fn start(&self) -> u16 {
+        self.start
+    }
 
-        let pos = self.from + idx;
-        let chunk = self.data[(pos / 8) as usize];
+    fn end(&self) -> u16 {
+        if self.is_leaf() {
+            KEY_SIZE as u16 * 8
+        } else {
+            u16::from(self.bytes[PROOF_PATH_LEN_POS])
+        }
+    }
+
+    fn bit(&self, idx: u16) -> ChildKind {
```

Comment:
It seems like `bit` and `common_prefix` methods can be moved to `BitRange` as default methods; they don't use `ProofPath` intrinsics. Is there some reason why they are not in `BitRange`?